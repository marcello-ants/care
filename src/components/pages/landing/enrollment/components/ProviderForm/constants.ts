export const PROVIDER_SERVICES_DESKTOP = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const PROVIDER_SERVICES_MOBILE = [
  {
    description: 'Child care',
    label: 'Child care',
    value: 'CHILD_CARE',
  },
  {
    description: 'Senior care',
    label: 'Senior care',
    value: 'SENIOR_CARE',
  },
  {
    description: 'Pet care',
    label: 'Pet care',
    value: 'PET_CARE',
  },
  {
    description: 'Tutoring',
    label: 'Tutoring',
    value: 'TUTORING',
  },
  {
    description: 'Special needs care',
    label: 'Special needs care',
    value: 'SPECIAL_NEEDS',
  },
  {
    description: 'Housekeeping',
    label: 'Housekeeping',
    value: 'HOUSEKEEPING',
  },
  {
    description: 'Errands & odd jobs',
    label: 'Errands & odd jobs',
    value: 'GENERIC',
  },
];

export const DROPOFF_URLS = {
  CHILD_CARE: '/swf/ccOptimizedFlow',
  SENIOR_CARE: '/app/enrollment/provider/sc ',
  PET_CARE: '/swf/pcOptimizedFlow',
  HOUSEKEEPING: '/swf/hkOptimizedFlow',
  TUTORING: '/swf/tuOptimizedFlow',
  SPECIAL_NEEDS: '/swf/snOptimizedFlow',
  GENERIC: '/swf/cgOptimizedFlow',
  AU_PAIR: '',
  CARE_GIGS: '',
};
