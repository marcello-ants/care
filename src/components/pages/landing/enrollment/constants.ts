export const SITTER_FORM = {
  providerServices: {
    title: 'What Services do you offer?',
    id: 'providerService',
    options: [
      {
        value: 'CHILD_CARE',
        text: 'Babysitter',
      },
      {
        value: 'CHILD_CARE',
        text: 'Nanny',
      },
      {
        value: 'SPECIAL_NEEDS',
        text: 'Special Needs Provider',
      },
      {
        value: 'TUTORING',
        text: 'Tutor',
      },
      {
        value: 'TUTORING',
        text: 'Private Lesson Instructor',
      },
      {
        value: 'SENIOR_CARE',
        text: 'Senior Care Provider',
      },
      {
        value: 'PET_CARE',
        text: 'Pet Care Provider',
      },
      {
        value: 'HOUSEKEEPING',
        text: 'Housekeeper',
      },
      {
        value: 'GENERIC',
        text: 'Errands &amp; Odd Jobs Provider',
      },
    ],
  },
  providerEmail: {
    id: 'email',
    placeholder: 'Email',
  },
  providerPassword: {
    id: 'password',
    placeholder: 'Password',
  },
  providerZip: {
    id: 'providerZip',
    placeholder: 'ZIP',
  },
};

export const GENERIC_FORM = {
  howDidYouHearAboutUs: {
    title: 'How did you hear about us?',
    id: 'howDidYouHearAboutUs',
    options: [
      {
        value: 'FACEBOOK',
        text: 'Facebook',
      },
      {
        value: 'BANNER_AD',
        text: 'Banner Ad',
      },
      {
        value: 'TV_AD',
        text: 'TV Ad',
      },
      {
        value: 'OTHER_SOCIAL_MEDIA',
        text: 'Other Social Media (Twitter, Instagram, Pinterest, LinkedIn)',
      },
      {
        value: 'PRESS_COVERAGE',
        text: 'Press Coverage (News, Magazine, Blog)',
      },
      {
        value: 'FRIENDS_OR_FAMILY',
        text: 'Friends or Family',
      },
      {
        value: 'ONLINE_VIDEO',
        text: 'Online Video (Hulu, Youtube)',
      },
      {
        value: 'SEARCH_ENGINE',
        text: 'Search Engine (Google, Bing)',
      },
      {
        value: 'PARENTING_GROUP',
        text: 'Parenting Group or Forum',
      },
      {
        value: 'OTHER',
        text: 'Other',
      },
    ],
  },
};
