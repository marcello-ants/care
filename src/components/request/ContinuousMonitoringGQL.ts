import { gql } from '@apollo/client';

export const GET_CONTINUOUS_MONITORING_ENROLLMENT_STATUS = gql`
  query ContinuousMonitoringEnrollmentStatus($memberId: ID!) {
    continuousMonitoringEnrollmentStatus(memberId: $memberId) {
      ... on ContinuousMonitoringEnrollmentStatusSuccess {
        needsConsent
        consentMessagesList {
          consentMessageId
          message
        }
      }
    }
  }
`;

export const ENROLL_IN_CONTINUOUS_MONITORING = gql`
  mutation EnrollInContinuousMonitoring($input: EnrollInContinuousMonitoringInput!) {
    enrollInContinuousMonitoring(input: $input) {
      ... on EnrollInContinuousMonitoringSuccess {
        success
      }
    }
  }
`;
