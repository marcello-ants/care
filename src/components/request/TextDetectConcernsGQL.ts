import { gql } from '@apollo/client';

// apollo client breaks if this query uses union type inside fragments and silently drops detectedConcerns from results
const TEXT_DETECT_CONCERNS = gql`
  query textAnalyzerDetectConcerns(
    $bio: String!
    $headline: String!
    $textCreator: TextAnalyzerDetectConcernsCreator!
    $validationId: String!
  ) {
    bio: textAnalyzerDetectConcerns(
      textToAnalyze: { text: $bio }
      textCreator: $textCreator
      validationId: $validationId
    ) {
      ... on TextAnalyzerDetectConcernsSuccess {
        __typename
        detectedConcerns {
          endIndex
          startIndex
          text
          type
        }
        validationId
      }
      ... on TextAnalyzerDetectConcernsError {
        __typename
        message
      }
    }
    headline: textAnalyzerDetectConcerns(
      textToAnalyze: { text: $headline }
      textCreator: $textCreator
      validationId: $validationId
    ) {
      ... on TextAnalyzerDetectConcernsSuccess {
        __typename
        detectedConcerns {
          endIndex
          startIndex
          text
          type
        }
        validationId
      }
      ... on TextAnalyzerDetectConcernsError {
        __typename
        message
      }
    }
  }
`;

export { TEXT_DETECT_CONCERNS };
