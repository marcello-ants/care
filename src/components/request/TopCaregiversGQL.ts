import { gql } from '@apollo/client';

const GET_TOP_CAREGIVERS = gql`
  query getTopCaregivers(
    $zipcode: String!
    $serviceID: ServiceType!
    $numResults: Int!
    $hourlyRate: JobRateInput
    $hasCareCheck: Boolean
    $qualities: [SeniorCareProviderQuality!]
    $services: [SeniorCareServiceProvidedType!]
    $maxDistanceFromSeeker: DistanceInput
    $hasApprovedActivePhoto: Boolean
    $enableFuzzySearch: Boolean
    $daysSinceLastActive: Int
    $isActiveAccount: Boolean
    $minimumAvgReviewRating: Float
    $fullTime: Boolean
  ) {
    topCaregivers: getTopCaregivers(
      zipcode: $zipcode
      numResults: $numResults
      serviceType: $serviceID
      hourlyRate: $hourlyRate
      hasCareCheck: $hasCareCheck
      qualities: $qualities
      services: $services
      maxDistanceFromSeeker: $maxDistanceFromSeeker
      hasApprovedActivePhoto: $hasApprovedActivePhoto
      enableFuzzySearch: $enableFuzzySearch
      daysSinceLastActive: $daysSinceLastActive
      isActiveAccount: $isActiveAccount
      minimumAvgReviewRating: $minimumAvgReviewRating
      fullTime: $fullTime
    ) {
      caregiver {
        profileURL
        avgReviewRating
        numberOfReviews
        yearsOfExperience
        hasCareCheck
        responseTime
        member {
          address {
            city
            state
          }
          displayName
          firstName
          id
          imageURL
          legacyId
        }
        profiles {
          commonCaregiverProfile {
            id
          }
          childCareCaregiverProfile {
            bio {
              experienceSummary
            }
            qualities {
              certifiedNursingAssistant
              certifiedRegisterNurse
              certifiedTeacher
              childDevelopmentAssociate
              comfortableWithPets
              cprTrained
              crn
              doesNotSmoke
              doula
              earlyChildDevelopmentCoursework
              earlyChildhoodEducation
              firstAidTraining
              nafccCertified
              ownTransportation
              specialNeedsCare
              trustlineCertifiedCalifornia
            }
            supportedServices {
              carpooling
              craftAssistance
              errands
              groceryShopping
              laundryAssistance
              lightHousekeeping
              mealPreparation
              swimmingSupervision
              travel
            }
            payRange {
              hourlyRateFrom {
                amount
                currencyCode
              }
              hourlyRateTo {
                amount
                currencyCode
              }
            }
          }
        }
        seniorCareProviderProfile {
          bio
          services
          payRange {
            hourlyRateFrom {
              amount
              currencyCode
            }
            hourlyRateTo {
              amount
              currencyCode
            }
          }
          qualities
        }
        signUpDate
      }
      distanceFromRequestZip {
        unit
        value
      }
    }
  }
`;

export default GET_TOP_CAREGIVERS;
