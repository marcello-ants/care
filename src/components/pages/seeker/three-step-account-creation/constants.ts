import * as yup from 'yup';
import { ServiceType } from '@/__generated__/globalTypes';
import { VerticalsAbbreviation } from '@/constants';

export const validationSchema = yup.object({
  firstName: yup
    .string()
    .min(2, 'First name needs to have at least two letters.')
    .required('First name is required'),
  lastName: yup
    .string()
    .min(2, 'Last name needs to have at least two letters.')
    .required('Last name is required'),
});

export const SERVICE_TYPE_BY_VERTICAL: { [key: string]: ServiceType } = {
  HK: ServiceType.HOUSEKEEPING,
  PC: ServiceType.PET_CARE,
  TU: ServiceType.TUTORING,
  CC: ServiceType.CHILD_CARE,
  DC: ServiceType.CHILD_CARE,
  IB: ServiceType.CHILD_CARE,
  SC: ServiceType.SENIOR_CARE,
};

export const ERROR_CREATING_ACCOUNT = 'An error occurred creating your account, please try again.';

export const FACEBOOK_VERTICALS_ALLOWANCE = ['CC'];
export const HDYHAU_VERTICALS_ALLOWANCE: VerticalsAbbreviation[] = ['HK', 'TU', 'PC', 'DC'];
