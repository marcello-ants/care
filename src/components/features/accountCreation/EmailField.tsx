import React, { ReactNode, useCallback, useEffect, useRef, useState } from 'react';
import { useApolloClient } from '@apollo/client';
import { useField, useFormikContext } from 'formik';
import { debounce } from 'lodash-es';
import * as yup from 'yup';
import { InlineTextField } from '@care/react-component-lib';
import { TextFieldProps } from '@material-ui/core';
import {
  validateMemberEmail,
  validateMemberEmailVariables,
} from '@/__generated__/validateMemberEmail';
import { VALIDATE_MEMBER_EMAIL } from '@/components/request/GQL';
import { SKIP_AUTH_CONTEXT_KEY } from '@/constants';
import { AnalyticsHelper } from '@/utilities/analyticsHelper';

const MEMBER_EMAIL_ALREADY_REGISTERED = 'MemberEmailAlreadyRegistered';

const emailSchema = yup.string().email();

type EmailFieldProps = {
  id: string;
  name: string;
  label: string;
  czenGeneralLoginUrl: string;
  showErrorWhenTouched?: boolean;
  classes?: TextFieldProps['classes'];
};

function EmailField(props: EmailFieldProps) {
  const ref = useRef<{ email?: string; errorMessage?: string }>({});
  const { id, name, label, czenGeneralLoginUrl, showErrorWhenTouched, classes } = props;
  const apolloClient = useApolloClient();

  useEffect(() => {
    if (ref.current.errorMessage?.startsWith(MEMBER_EMAIL_ALREADY_REGISTERED)) {
      AnalyticsHelper.logEvent({
        name: 'Email Already Registered',
      });
    }
  }, [ref.current.errorMessage]);

  const debouncedValidation = useCallback(
    debounce(async (email: string, resolve: (msg: string | void) => void) => {
      let responseError;
      try {
        const { data } = await apolloClient.query<
          validateMemberEmail,
          validateMemberEmailVariables
        >({
          query: VALIDATE_MEMBER_EMAIL,
          variables: { email },
          context: { [SKIP_AUTH_CONTEXT_KEY]: true },
        });
        responseError = data?.validateMemberEmail.errors?.[0];
      } catch (e) {
        responseError = null;
      }

      let errorMessage = responseError ? responseError.message || 'Validation Failed' : undefined;

      if (responseError?.__typename === MEMBER_EMAIL_ALREADY_REGISTERED) {
        // we want to inject a log in link into this message, but formik expects a string
        // to get around this, prefix the message with the typename so we can identify and remove it on render
        errorMessage = `${MEMBER_EMAIL_ALREADY_REGISTERED}${errorMessage}`;
      }

      resolve(errorMessage);
      ref.current = {
        email,
        errorMessage,
      };
    }, 500),
    [apolloClient]
  );

  const validate = useCallback(
    async (value: string) => {
      if (ref.current.email === value) {
        return ref.current.errorMessage;
      }

      const formatIsValid = await emailSchema.isValid(value);
      if (!formatIsValid) {
        return 'Please enter a valid email.';
      }

      return new Promise<string | void>((resolve) => {
        debouncedValidation(value, resolve);
      });
    },
    [debouncedValidation]
  );

  const [field, meta] = useField({ name, validate });
  const [isFocused, setIsFocused] = useState(false);
  const { isValidating } = useFormikContext();

  // server validation errors will be immediately shown, but client-side (schema/regex) validations show when the field is blurred
  const showError = Boolean(
    meta.error &&
      !isValidating &&
      (showErrorWhenTouched ? meta.touched : field.value) &&
      (!isFocused || emailSchema.isValidSync(field.value))
  );

  let helperText: ReactNode;
  if (showError && meta.error) {
    helperText = meta.error;
    if (meta.error.startsWith(MEMBER_EMAIL_ALREADY_REGISTERED)) {
      helperText = (
        <span>
          {meta.error.replace(MEMBER_EMAIL_ALREADY_REGISTERED, '')}{' '}
          <a href={czenGeneralLoginUrl} style={{ textDecoration: 'none' }}>
            Log in?
          </a>
        </span>
      );
    }
  }

  function handleFocus() {
    setIsFocused(true);
  }

  function handleBlur(e: React.FocusEvent) {
    setIsFocused(false);
    field.onBlur(e);
  }

  return (
    <InlineTextField
      type="email"
      id={id}
      name={field.name}
      label={label}
      value={field.value}
      error={showError}
      helperText={helperText}
      onChange={field.onChange}
      onBlur={handleBlur}
      onFocus={handleFocus}
      classes={classes}
    />
  );
}

EmailField.defaultProps = {
  showErrorWhenTouched: false,
  classes: undefined,
};

export default EmailField;
