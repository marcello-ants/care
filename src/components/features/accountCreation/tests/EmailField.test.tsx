import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { Formik } from 'formik';
import { VALIDATE_MEMBER_EMAIL } from '@/components/request/GQL';
// eslint-disable-next-line camelcase
import { validateMemberEmail_validateMemberEmail_errors } from '@/__generated__/validateMemberEmail';
import { AnalyticsHelper } from '@/utilities/analyticsHelper';
import { SEEKER_ROUTES } from '@/constants';
import EmailField from '../EmailField';

jest.mock('next/config', () => () => {
  return {
    publicRuntimeConfig: {
      CZEN_GENERAL_LOGIN: 'https://www.care.com/login',
    },
  };
});

jest.mock('@/utilities/analyticsHelper', () => ({
  ...jest.requireActual('@/utilities/analyticsHelper'),
  AnalyticsHelper: {
    logEvent: jest.fn(),
  },
}));

const emailValidationMock = (
  email: string,
  // eslint-disable-next-line camelcase
  error?: validateMemberEmail_validateMemberEmail_errors
) => ({
  request: {
    query: VALIDATE_MEMBER_EMAIL,
    variables: {
      email,
    },
  },
  result: {
    data: {
      validateMemberEmail: { errors: [error] },
    },
  },
});

let emailField: HTMLElement;

function renderComponent(mocks?: ReadonlyArray<MockedResponse>, props?: { [key: string]: any }) {
  const view = render(
    <MockedProvider mocks={mocks} addTypename={false}>
      <Formik
        initialValues={{
          email: '',
        }}
        validate={(values) => {
          const errors = {};

          if (!values.email) {
            // @ts-ignore
            errors.email = 'Email is required.';
          }

          return errors;
        }}
        onSubmit={() => {}}
        validateOnMount>
        <EmailField
          id="email"
          name="email"
          label="Email Address"
          czenGeneralLoginUrl="https://www.care.com/login"
          {...props}
        />
      </Formik>
    </MockedProvider>
  );

  emailField = screen.getByLabelText('Email Address');

  return view;
}

describe('EmailField', () => {
  it('matches snapshot', () => {
    const view = renderComponent();
    expect(view.asFragment()).toMatchSnapshot();
  });

  it('should display a validation error when an invalid email is entered', async () => {
    renderComponent();

    fireEvent.change(emailField, { target: { value: 'tim' } });

    const validationMsg = await screen.findByText('Please enter a valid email.');
    expect(validationMsg).toBeVisible();
  });

  it('should not display the invalid email message when the field is focused', async () => {
    renderComponent();

    fireEvent.change(emailField, { target: { value: 'tim' } });

    await screen.findByText('Please enter a valid email.');

    fireEvent.focus(emailField);

    await waitFor(() =>
      expect(screen.queryByText('Please enter a valid email.')).not.toBeInTheDocument()
    );
  });

  it('should display the invalid email message when the field is blurred', async () => {
    renderComponent();

    fireEvent.change(emailField, { target: { value: 'tim' } });
    fireEvent.focus(emailField);

    await waitFor(() =>
      expect(screen.queryByText('Please enter a valid email.')).not.toBeInTheDocument()
    );

    fireEvent.blur(emailField);

    expect(await screen.findByText('Please enter a valid email.')).toBeVisible();
  });

  it('should display a login link when the entered email is already registered and call Amplitude tracking', async () => {
    Object.defineProperty(window, 'location', {
      get() {
        return {
          href: `http://localhost${SEEKER_ROUTES.ACCOUNT_CREATION_NAME}`,
          pathname: SEEKER_ROUTES.ACCOUNT_CREATION_NAME,
        };
      },
    });
    renderComponent([
      emailValidationMock('tim@care.com', {
        __typename: 'MemberEmailAlreadyRegistered',
        message: 'This email is already registered.',
      }),
    ]);

    fireEvent.change(emailField, { target: { value: 'tim@care.com' } });

    const validationMsg = await screen.findByText('This email is already registered.');
    expect(validationMsg).toBeVisible();

    const loginLink = screen.getByRole('link', { name: 'Log in?' });
    expect(loginLink).toHaveAttribute('href', 'https://www.care.com/login');

    expect(AnalyticsHelper.logEvent).toHaveBeenCalledWith({
      name: 'Email Already Registered',
    });
  });

  it('should display "validation failed" when an unexpected response is returned', async () => {
    renderComponent([
      emailValidationMock('tim@care.com', {
        __typename: 'MemberEmailInvalid',
        message: '',
      }),
    ]);

    fireEvent.change(emailField, { target: { value: 'tim@care.com' } });

    const validationMsg = await screen.findByText('Validation Failed');
    expect(validationMsg).toBeVisible();
  });

  it('should display email is empty message when the field is blurred and empty', async () => {
    renderComponent([], { showErrorWhenTouched: true });

    fireEvent.focus(emailField);
    fireEvent.blur(emailField);

    expect(await screen.findByText('Email is required.')).toBeVisible();
  });
});
