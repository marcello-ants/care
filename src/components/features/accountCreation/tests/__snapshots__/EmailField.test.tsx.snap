// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailField matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root MuiTextField-root makeStyles-inline-1"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink="false"
      for="email"
      id="email-label"
    >
      Email Address
    </label>
    <div
      class="MuiInputBase-root MuiInput-root MuiInputBase-formControl MuiInput-formControl"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiInput-input"
        id="email"
        name="email"
        type="email"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
